pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,913 @@
+				shareInfo, err := recon.assignInstanceToShareInfo(shareInfo, instanceURI)
+				if err != nil {
+					klog.Errorf("cannot update instanceHandle to shareInfo %q: %v", shareName, err)
+					continue
+				}
+				shareInfos[shareName] = shareInfo
+			} else if shareInfo.Status.InstanceHandle != instanceURI {
+				// This case should be rare, however, if there's a race or crash between shareInfo update and instanceInfo update, the 2 way pointer may not match.
+				klog.Warningf("instanceInfo %q has share %q but its shareInfo points to instance %q", instanceInfo.Name, shareName, shareInfo.Status.InstanceHandle)
+
+				// if the share is already marked for deletion, don't try to add it to the instanceInfo it points to.
+				if shareInfo.DeletionTimestamp == nil {
+					klog.Infof("Deletion Timestamp not set on shareInfo %q, trying to update instanceInfo %q", shareName, shareInfo.Status.InstanceHandle)
+					actualAssigned, ok := instanceInfos[shareInfo.Status.InstanceHandle]
+					var err error
+					if !ok {
+						klog.Errorf("share %q is assigned to instance %q but instanceInfo does not exist. Trying to create one", shareName, shareInfo.Status.InstanceHandle)
+						actualAssigned, err = recon.generateInstanceInfo(shareInfo.Status.InstanceHandle, shareInfo.Spec.InstancePoolTag)
+						if err != nil {
+							klog.Errorf("failed to create instanceInfo %q: %v", shareInfo.Status.InstanceHandle, err)
+							continue
+						}
+					}
+					actualAssigned, err = recon.assignShareToInstanceInfo(actualAssigned, shareName)
+					if err != nil {
+						klog.Errorf("failed to assign share %q to instanceInfo %q: %v", shareName, shareInfo.Status.InstanceHandle, err)
+						continue
+					}
+
+					instanceInfos[shareInfo.Status.InstanceHandle] = actualAssigned
+				}
+
+				instanceInfoClone, updated = recon.removeShareFromInstanceInfo(instanceInfoClone, shareName)
+			}
+		}
+
+		var err error
+		if updated {
+			klog.Infof("instanceInfo %q has updated share assignment, trying to update object", instanceInfoClone.Name)
+			instanceInfoClone, err = recon.updateInstanceInfoStatus(context.TODO(), instanceInfoClone)
+			if err != nil {
+				klog.Errorf("failed to update status subresource of instanceInfo %q: %v", instanceInfo.Name, err)
+				continue
+			}
+			instanceInfos[instanceURI] = instanceInfoClone
+		}
+	}
+}
+
+// assignSharesToEligibleOrNewInstances assigns shares that are not already assigned to eligible instances.
+// If there're no eligible instances, generate a new one.
+func (recon *multishareReconciler) assignSharesToEligibleOrNewInstances(shareInfos map[string]*v1alpha1.ShareInfo, instanceInfos map[string]*v1alpha1.InstanceInfo) {
+	for _, shareInfo := range shareInfos {
+		if shareInfo.Status == nil || shareInfo.Status.InstanceHandle == "" {
+
+			// if share is up to delete, don't assign it even if it's not yet assigned.
+			// This situation should not happen because DeleteVolume shouldn't be called until CreateVolume already succeeded
```

Comment:
We need to write a test case for this. invoke delete for CSI create completes

Reply:
before CSI create completes right? a guitar test might work, I think the main safeguard is that CSI provisioner will make sure that `deleteVolume` is called after `createVolume` return success.

Reply:
correct. a test case would ensure there is no regression